File: ios/Classes/Scanner Core/Scan Filters/CardHolderNameFilter.swift
//
//  CardHolderNameFilter.swift
//  card_scanner
//
//  Created by Abhinav Kumar on 12/02/21.
//

import Foundation
import MLKitTextRecognition

class CardHolderNameFilter : ScanFilter {
    private var cardHolderRegex: NSRegularExpression =  try! NSRegularExpression(pattern: CardScannerRegexps.cardHolderName, options: .anchorsMatchLines)
    var maxBlocksBelowCardNumberToSearchForName = 4
    
    func filter() -> ScanFilterResult? {
        guard scannerOptions.scanCardHolderName else { return nil }
        guard cardNumberScanResult.cardNumber.isNotEmpty else { return nil }
        
        // Search from card number block and below [_maxBlocksBelowCardNumberToSearchForName] blocks
        let minTextBlockIndexToSearchName = max(cardNumberScanResult.textBlockIndex - (scannerOptions.possibleCardHolderNamePositions.contains(CardHolderNameScanPositions.aboveCardNumber.rawValue) ? 1 : 0) , 0)
        
        let maxTextBlockIndexToSearchName =
            min(cardNumberScanResult.textBlockIndex +
                    (scannerOptions.possibleCardHolderNamePositions.contains(CardHolderNameScanPositions.belowCardNumber.rawValue)  ? maxBlocksBelowCardNumberToSearchForName : 0), visionText.blocks.count - 1)
        
        for index in minTextBlockIndexToSearchName...maxTextBlockIndexToSearchName {
            let block = visionText.blocks[index]
            let transformedBlockText = transformBlockText(blockText: block.text)
            let nsString = transformedBlockText as NSString
            if (!(cardHolderRegex.numberOfMatches(in: transformedBlockText, range: NSRange(location: 0, length: transformedBlockText.count)) > 0)) {
                continue
            }
            
            let firstMatch = cardHolderRegex.firstMatch(in: transformedBlockText, range: NSRange(location: 0, length: transformedBlockText.count))
            
            let cardHolderName = nsString.substring(with: firstMatch?.range ?? NSRange(location: 0, length: 0)).trimmingCharacters(in: .whitespacesAndNewlines) as String
            
            
            if (isValidName(cardHolder: cardHolderName)) {
                return CardHolderNameScanResult(
                    visionText: visionText, textBlockIndex: index, textBlock: block, cardHolderName: cardHolderName)
            }
        }
        return nil
    }
    
    var visionText: Text
    var scannerOptions: CardScannerOptions
    private var cardNumberScanResult: CardNumberScanResult
    
    init(visionText: Text, scannerOptions: CardScannerOptions, cardNumberScanResult: CardNumberScanResult) {
        self.visionText = visionText
        self.scannerOptions = scannerOptions
        self.cardNumberScanResult = cardNumberScanResult
    }
    
    func isValidName(cardHolder: String) -> Bool {
        if (cardHolder.count < 3 || cardHolder.count > scannerOptions.maxCardHolderNameLength) {
            debugLog("maxCardHolderName length = \(scannerOptions.maxCardHolderNameLength)", scannerOptions: scannerOptions)
            return false
        }
        
        if (cardHolder.hasPrefix("valid from") || cardHolder.hasPrefix("valid thru")) { return false }
        if (cardHolder.hasSuffix("valid from") || cardHolder.hasSuffix("valid thru")) { return false }
        var defaultBlackListedWords = CardHolderNameConstants.defaultBlackListedWords
        defaultBlackListedWords.append(contentsOf: scannerOptions.cardHolderNameBlackListedWords)
        
        if (defaultBlackListedWords.contains(cardHolder.lowercased())) {
            return false
        }
        return true
    }
    
    func transformBlockText(blockText: String) -> String {
        return blockText.replacingOccurrences(of: "c", with: "C")
            .replacingOccurrences(of: "o", with: "O")
            .replacingOccurrences(of: "p", with: "P")
            .replacingOccurrences(of: "v", with: "V")
            .replacingOccurrences(of: "w", with: "W")
    }
}

File: android/src/main/java/com/nateshmbhat/card_scanner/scanner_core/scan_filters/CardHolderNameFilter.kt
package com.nateshmbhat.card_scanner.scanner_core.scan_filters

import android.util.Log
import com.google.mlkit.vision.text.Text
import com.nateshmbhat.card_scanner.logger.debugLog
import com.nateshmbhat.card_scanner.scanner_core.constants.CardHolderNameConstants
import com.nateshmbhat.card_scanner.scanner_core.constants.CardScannerRegexps
import com.nateshmbhat.card_scanner.scanner_core.models.*
import java.util.*
import kotlin.math.max
import kotlin.math.min

class CardHolderNameFilter(visionText: Text, scannerOptions: CardScannerOptions, private val cardNumberScanResult: CardNumberScanResult) : ScanFilter(visionText, scannerOptions) {
  private val cardHolderRegex: Regex = Regex(CardScannerRegexps.cardHolderName, RegexOption.MULTILINE)
  val _maxBlocksBelowCardNumberToSearchForName = 4;

  override fun filter(): CardHolderNameScanResult? {
    if (!scannerOptions.scanCardHolderName) return null;
    if (cardNumberScanResult.cardNumber.isEmpty()) return null;

    ///Search from card number block and below [_maxBlocksBelowCardNumberToSearchForName] blocks
    val minTextBlockIndexToSearchName = max(cardNumberScanResult.textBlockIndex -
            (if (scannerOptions.possibleCardHolderNamePositions.contains(CardHolderNameScanPositions.aboveCardNumber.value)) 1 else 0), 0)
    val maxTextBlockIndexToSearchName =
            min(cardNumberScanResult.textBlockIndex +
                    (if (scannerOptions.possibleCardHolderNamePositions.contains((CardHolderNameScanPositions.belowCardNumber.value))) _maxBlocksBelowCardNumberToSearchForName else 0), visionText.textBlocks.size - 1);

    for (index in minTextBlockIndexToSearchName..maxTextBlockIndexToSearchName) {
      val block = visionText.textBlocks[index]
      val transformedBlockText = transformBlockText(block.text)
      if (!cardHolderRegex.containsMatchIn(transformedBlockText)) continue;
      val cardHolderName = cardHolderRegex.find(transformedBlockText)!!.value.trim()
      if (isValidName(cardHolderName)) {
        return CardHolderNameScanResult(
                textBlockIndex = index, textBlock = block, cardHolderName = cardHolderName, visionText = visionText);
      }
    }
    return null;
  }

  private fun isValidName(cardHolder: String): Boolean {
    if (cardHolder.length < 3 || cardHolder.length > scannerOptions.maxCardHolderNameLength) {
      debugLog("maxCardHolderName length = " + scannerOptions.maxCardHolderNameLength, scannerOptions);
      return false
    };
    if (cardHolder.startsWith("valid from") || cardHolder.startsWith("valid thru")) return false;
    if (cardHolder.endsWith("valid from") || cardHolder.endsWith("valid thru")) return false;
    if (CardHolderNameConstants.defaultBlackListedWords
                    .union(scannerOptions.cardHolderNameBlackListedWords.toSet())
                    .contains(cardHolder.toLowerCase(Locale.ENGLISH))) {
      return false;
    }
    return true;
  }

  private fun transformBlockText(blockText: String): String {
    return blockText.replace('c', 'C')
            .replace('o', 'O')
            .replace('p', 'P')
            .replace('v', 'V')
            .replace('w', 'W')
  }
}